### DISCUSSION TITLE
Mehrsprachige GUI?

### DISCUSSION THREAD

**Discussion by manoloxxl** [Zefau/ioBroker.jarvis#2724] [General]
Hi,

gibt es einen einfachen Weg einen Sprach-Selektor in die Jarvis GUI einzubauen um es dem Bediener zu ermöglichen zwischen z. B. deutschen und englischen Labels/Überschriften/JSON Elementen zu wechseln?

Grüße

**Comment by Zefau** [comment]
Sprache kann über `jarvis.0.settings` dynamisch angepasst werden

**Comment by manoloxxl** [comment]
Und leider muss ich auch hier fragen, hab in der Doku nichts gefunden.

Die Sprache kann ich umstellen, aber wie hinterlege ich Übersetzungen bei Gerätenamen, Buttons und Überschriften damit diese bei Englisch auch was anderes anzeigen als bei Deutsch?

**Reply by mcuiobroker** [reply]
Ist noch nicht vorgesehen (Feature Request), User-Abhängige Inhalte, abhängig von der eingestellten Sprache zu machen. 
Dafür müsstest du einen eigenen Sprache-DP für dich anlegen und den nutzen, um diese Werte entsprechend umzuschalten. 
Dies geht aber auch noch nicht für alle Elemente wie Gerätenamen usw. nur für einige Elemente mit JSON vorhanden. 
![image](https://github.com/user-attachments/assets/193b002b-93d9-4e0a-82e8-b3caa36ce560)

**Reply by manoloxxl** [reply]
Okay, danke dir. Dann muss ich mal überlegen ob ich das evtl. "einfach" mit einer zweiten Instanz löse.

**Reply by manoloxxl** [reply]
Hab mich mal dran versucht. Zweite Instanz fällt bei mir flach weil ich dann ständig das iPad umkonfigurieren müsste welches im geführten Modus läuft. Dafür brauche ich es dann doch zu selten Mehrsprachig.

Eine weitere Seite die der User selber anwählen kann würde als Alternative auch gehen, da müsste man aber alle Geräte, Widgets und Designs doppelt anlegen. Auch irgendwie uncool.


Generell geht das mit dem JSON gut in Geräten, wenn ich dann allerdings im Widget was setze wird mir nur mein JSON in der GUI ausgeworfen. Ich tippe mal auf einen bekannten Fehler in meiner Version und werde das mal fortsetzen wenn ich weiter updaten kann auf die aktuellste Version.

![translate](https://github.com/user-attachments/assets/33bbf2f3-fb64-470d-897a-f34ea42a5098)

Falls jemand ähnliches vorhat, es geht mit einem simplen JSON im Label Feld und einem DP. Ich hab es mit einem boolean gemacht weil ich nur auf EN umschalten will. Es gehen aber natürlich auch mehr Sprachen wenn man eine Liste nimmt und im JSON anders abfragt:

```
{
  "default": "Deutscher Text",
  "{0_userdata.0.Jarvis.show_jarvis_english}=true": "Englischer Text"
}
```

**Comment by manoloxxl** [comment]
Inzwischen bin ich weiter. Viel geht über das JSON, anderes über Scripte. Man bekommt die GUI relativ dynamisch übersetzt.

Ich scheitere aktuell noch daran dynamisch in der **Display Variant** zu arbeiten. Bisher hatte ich für Rauchmelder folgendes JSON:

`
{"true":"ALARM","false":"keine Meldung"}
`

Jetzt würde ich die Begriffe gerne, abhängig von einem Datenpunkt, übersetzen. Ich schaffe es aber nicht. Wobei ich nicht weiß ob ich es falsch verschachtel oder ob es einfach nicht geht. Ich würde es so bauen:

```
{
  "true": "ALARM",
  "false": {
    "{0_userdata.0.Jarvis.en-Gast}=false": "keine Meldung",
    "{0_userdata.0.Jarvis.en-Gast}=true": "no message"
  }
}
```

Dabei kommt in der GUI dann aber nur das JSON als Plaintext an.

Hat dazu jemand einen Tipp?

**Comment by mcuiobroker** [comment]
Steht was im Browser-LOG?

**Reply by manoloxxl** [reply]
Leider Nein:

![image](https://github.com/user-attachments/assets/c4d13408-7eb5-4966-b9ea-be5929e34515)

**Comment by mcuiobroker** [comment]
Sehr wahrscheinlich wird ein Objekt im `value` da noch nicht unterstützt?
Evtl. muss man es mit Datenpunkt-Eigenschaften lösen.

**Comment by manoloxxl** [comment]
Ich finde das Verhalten halt etwas verwirrend. Wenn ich zb eingebe:

```
{
  "{0_userdata.0.Jarvis.en-Gast}=false": "nicht en",
  "{0_userdata.0.Jarvis.en-Gast}=true": "EN"
}
```

Wird abhängig vom Datenpunkt der Text korrekt ausgegeben.

Ergänze ich dann meine Übersetzung abhängig vom Wert so:

```
{
  "{0_userdata.0.Jarvis.en-Gast}=false": "{"true":"ALARM","false":"keine Meldung"}",
  "{0_userdata.0.Jarvis.en-Gast}=true": "{"true":"ALARM","false":"no message"}"
}
```

Wird das Komma zwischen ""ALARM","false"" bemängelt. Baue ich es etwas um in:

```
{
  "{0_userdata.0.Jarvis.en-Gast}=false": {"true":"ALARM","false":"keine Meldung"},
  "{0_userdata.0.Jarvis.en-Gast}=true": {"true":"ALARM","false":"no message"}
}

```

Bekomme ich immer nur ein "false" in der GUI.

**Reply by mcuiobroker** [reply]
Mehrfach-Bindings (verschachtelt) werden halt nicht unterstützt, noch nicht?
Über Datenpunkt-Eigenschaften geht es auch nicht. Teste noch.

**Comment by mcuiobroker** [comment]
Und so?
```
{
  "{0_userdata.0.Jarvis.en-Gast}=false": "{"{val}=true":"ALARM","{val}=false":"keine Meldung"}",
  "{0_userdata.0.Jarvis.en-Gast}=true": "{"{val}=true":"ALARM","{val}=false":"no message"}"
}
```

**Reply by mcuiobroker** [reply]
Mit Datenpunkt-Eigenschaften geht es nicht.

**Reply by manoloxxl** [reply]
> Und so?
> 
> ```
> {
>   "{0_userdata.0.Jarvis.en-Gast}=false": "{"{val}=true":"ALARM","{val}=false":"keine Meldung"}",
>   "{0_userdata.0.Jarvis.en-Gast}=true": "{"{val}=true":"ALARM","{val}=false":"no message"}"
> }
> ```

Da bekomme ich direkt einen Fehler im GUI Editor. Expected , instead of {

**Comment by mcuiobroker** [comment]
Evtl. kann man es mit scripts lösen?
Welche Elemente Label, value, usw müssen gesetzt werden?
Wieviele Elemente sind es?
Nur bei StateLists? Wo geht es schon und wo noch nicht?

**Reply by manoloxxl** [reply]
> Evtl. kann man es mit scripts lösen? Welche Elemente Label, value, usw müssen gesetzt werden? Wieviele Elemente sind es? Nur bei StateLists? Wo geht es schon und wo noch nicht?

Wäre jetzt auch mein nächster Ansatz. Ich habe ein paar Texte die ich gerne übersetzen möchte die sich auch wiederholen (geschlossen, offen, Luftfeuchtigkeit....). Sind dann eigentlich alles "Anzeigevarianten" mit zwei bis drei Werten. Auch mal DropDowns bei.

Geht um Rauchmelder, Fensterstatus. Überwiegend reine Displaytexte in StateLists. Teils auch CustomTextBody.

**Comment by mcuiobroker** [comment]
So?
```
{
  "{0_userdata.0.Jarvis.en-Gast}=false": {"{val}=true":"ALARM","{val}=false":"keine Meldung"},
  "{0_userdata.0.Jarvis.en-Gast}=true": {"{val}=true":"ALARM","{val}=false":"no message"}
}
```

Da waren zu viele " drin
![image](https://github.com/user-attachments/assets/fb344de5-a047-45a5-8a5f-5e8d6e2a9eb8)

**Reply by manoloxxl** [reply]
> {
>   "{0_userdata.0.Jarvis.en-Gast}=false": {"{val}=true":"ALARM","{val}=false":"keine Meldung"},
>   "{0_userdata.0.Jarvis.en-Gast}=true": {"{val}=true":"ALARM","{val}=false":"no message"}
> }

Ergibt eine Ausgabe in der GUI die so aussieht:

{"false==true":"ALARM","false==false":"no message"}

**Reply by mcuiobroker** [reply]
Ja, dann unterstützt er keine Objekte.
Was ist mit scripts?

**Comment by manoloxxl** [comment]
Genau, Scripte wären dann auch ein Ausweg. Sind ein paar StateList und CustomBodyText.

Was mir aber gerade aufgefallen ist, evtl. kann man auf die Logik aufspringen. Bei vordefinierten Geräten wir z. B. Fenstern füllt Jarvis die Anzeigevariante ja automatisch. Sichtbar im JSON (aber nicht editierbar) sind hier die Statesd open, closed usw.:

![image](https://github.com/user-attachments/assets/7e195b5e-f1ae-41d0-b1e7-6065ba6e7a85)

Stelle ich die Jarvis-Sprache auf Englisch, zeigt er die Werte auch in englisch mit open, close usw. an. Stelle ich auf deutsch, zeigt die GUI offen, geschlossen usw an.

Eine Idee ob man das irgendwie nutzen kann?

**Comment by manoloxxl** [comment]
Und hier nun das fast vorläufige Ergebnis. Vorläufig, weil ich an Grenzen gestoßen bin bei denen ich keine Lösung gefunden habe. Mir fehlen noch:

* Übersetzungen für CustomBodyTexte - habe ich keine Lösung gefunden. Kein JSON möglich
* Button Labels - habe ich keine Lösung gefunden. Ich habe einige Buttons nun durch Icons ersetzt (Pfeile z. B., aus "Starten / Stoppen" ist "Start / Stop" geworden)
* Anzeigevariante/DisplayVariant abhängig von der Sprache setzen > scheint nicht auf DP zugreifen zu können. Siehe diesen Thread hier
* Widgets abhängig zur Sprache anzeigen - ich habe eine HTML Tabelle. Das zugehörige Widget wollte ich duplizieren und dann, je nach Sprache, das eine ein oder ausblenden. Dabei wird jedoch das Grid durch das ausgeblendete nicht verändert. Ich hatte gehofft die Box rückt nach, hat aber nicht geklappt. Ich überlege nun den DP im ioBroker beim Sprachwechsel jeweils anzupassen, sprich: beim Wechsel der Sprache das DE oder das EN HTML in den DP schreiben.

Es sind also noch ein paar deutsche Texte drin wie z. B. bei den Rauchmeldern (Texte aus der Anzeige Variante), in DropDowns mit denen was gesteuert wird usw. 

Grundsätzlich schalte ich mit dem Switch in der Nav-Bar (Danke MCUiobroker dafür - [hier die Diskussion](https://github.com/Zefau/ioBroker.jarvis/discussions/2884) dazu) einen DP im iobroker. Dieser DP ist auch die Bedingung im JSON der Devices um entweder default DE anzuzeigen oder wenn true eben den englischen Text.

Der DP wird auch in den Scripten geprüft, mit denen ich die Labels der Widgets und Labels der PopUps setze:
devicepopuplabelchange.js
widgetlabelchangetxt.js

Wird der DP durch den Switch gesetzt schalte ich ebenso den Jarvis-Sprach-DP um: (jarvis.0.settings.language) von de-DE auf en-GB damit die automatischen Texte "x days ago" und einige States wie bei Fenstern (open/close) automatisch übersetzt werden. Ebenso schaltet dabei auch z. B. das Wetter und Date Widget die Sprache um.

Damit das aber zuverlässig greift musste ich einen[ GUI Reload](https://mcuiobroker.gitbook.io/jarvis-infos/jarvis-v3/besonderheiten-v3/scripts-3.1.x-pro/sonstige-skripte/reloadpage) einbauen. Der ist im Video zu sehen: 2 Sekunden nach umstellen des Sprachwählers lädt die Seite neu und zeigt alles in der korrekten Sprache an. Ich finde es nicht mal zwingend störend. Man schaltet schließlich nicht ständig hin und her.

Nachteile dieser Lösung:
...neben den nicht übersetzen Texten... ;)
1) Alle Geräte in der Instanz wechseln die Sprache wenn sie an einem Gerät umgestellt wird. Ich habe nur ein Gerät, ist mir also egal.
2) die Jarvis Admin GUI ist dann natürlich auch englisch. Da hält man sich im laufenden Betrieb aber ja eigentlich nicht drin auf
3) beim Testen läuft man in einen Fehler mit der Sprachauswahl weil diese im ggf. offenen Settingstab sich nicht umstellt. Es kann also hilfreich sein das Ergebnis dann ohne offenen Admin zu testen

![JarvisLanguageSwitch](https://github.com/user-attachments/assets/7a88c3ff-9863-4247-8de3-4ecdddc631a5)


Falls jemand Tipps zu dem Umlabeln der Buttons hat, gerne her damit. 
Die DisplayVariant wird vielleicht irgendwann kommen, dazu werde ich noch einen Issue einstellen.

Mir reicht das so, die wichtigsten Dinge sind jetzt in Englisch verfügbar.

Und im Nachgang hätte man das sicher auch stumpf durch Google Translate übersetzen können mit einer zweiten URL ;)

**Comment by mcuiobroker** [comment]
>Widgets abhängig zur Sprache anzeigen - ich habe eine HTML Tabelle. Das zugehörige Widget wollte ich duplizieren und dann, je nach Sprache, das eine ein oder ausblenden. Dabei wird jedoch das Grid durch das ausgeblendete nicht verändert. Ich hatte gehofft die Box rückt nach, hat aber nicht geklappt. Ich überlege nun den DP im ioBroker beim Sprachwechsel jeweils anzupassen, sprich: beim Wechsel der Sprache das DE oder das EN HTML in den DP schreiben.

Hab es nicht probiert, aber könnte mir vorstellen, dass man mit der Anzeige-Variante die Tabellen austauschen kann.
```
{"false":"{DP-von-Tabelle 1}","true":"{DP-von-Tabelle 2}"}
```
Module stateHTML mit Geräte-DP

**Reply by manoloxxl** [reply]
> > Widgets abhängig zur Sprache anzeigen - ich habe eine HTML Tabelle. Das zugehörige Widget wollte ich duplizieren und dann, je nach Sprache, das eine ein oder ausblenden. Dabei wird jedoch das Grid durch das ausgeblendete nicht verändert. Ich hatte gehofft die Box rückt nach, hat aber nicht geklappt. Ich überlege nun den DP im ioBroker beim Sprachwechsel jeweils anzupassen, sprich: beim Wechsel der Sprache das DE oder das EN HTML in den DP schreiben.
> 
> Hab es nicht probiert, aber könnte mir vorstellen, dass man mit der Anzeige-Variante die Tabellen austauschen kann.
> 
> ```
> {"false":"{DP-von-Tabelle 1}","true":"{DP-von-Tabelle 2}"}
> ```
> 
> Module stateHTML mit Geräte-DP

Hab ich ausprobiert mit:

`{"default":"{0_userdata.0.Jarvis.HTML_Telefonnummern}","{0_userdata.0.Jarvis.en-Gast}=true":"{0_userdata.0.Jarvis.HTML_Telefonnummern_en}"}
`
im Gerät.

In State (to display) bleibt das Feld dann leer (keine Meldung in der Console)
Bei Display Variant kommt immer nur die DE Variante.

**Reply by mcuiobroker** [reply]
Direkt ein Geräte DP für den Umschalt-DP anlegen und den nutzen
dann mit "false"... und "true" arbeiten

**Reply by manoloxxl** [reply]
> Direkt ein Geräte DP für den Umschalt-DP anlegen und den nutzen dann mit "false"... und "true" arbeiten

Das Gerät "EN-Gast" als StateHTML und dort mit Anzeigevariante 

{"false":"{DP-von-Tabelle 1}","true":"{DP-von-Tabelle 2}"}

arbeiten?

**Reply by mcuiobroker** [reply]
ja, Geräte-DP evtl extra neu anlegen

**Reply by manoloxxl** [reply]
Endet leider auch in einer leeren Anzeige :(

Daher das mit dem Kopieren klappt, schaue ich mir das mit dem Device später noch mal an.

**Reply by mcuiobroker** [reply]
Dann musst du die Umschaltung in ioBroker machen
EN-Umschaltung-DP 
HTML-DP
html1-DP
html2-DP
```
if(EN){
   setState(html-DP,html1-DP,true)
}else{
   setState(html-DP,html2-DP,true)
}
```
